apisix/plugins/auth-hook.lua

```diff
@@ -0,0 +1,337 @@
+    myheaders["Content-Type"] = "application/json; charset=utf-8"
+    local auth_hook_params = {
+        ssl_verify = false,
+        method = config.auth_hook_method,
+        headers = myheaders,
+    };
+    local url = config.auth_hook_uri;
+    -- Keepalive options.
+    if config.hook_keepalive then
+        auth_hook_params.keepalive_timeout = config.hook_keepalive_timeout
+        auth_hook_params.keepalive_pool = config.hook_keepalive_pool
+    else
+        auth_hook_params.keepalive = config.hook_keepalive
+    end
+    url = config.auth_hook_uri .. "?" .. ngx.encode_args(args)
+    if config.auth_hook_method == 'POST' then
+        auth_hook_params.body = nil
+    else
+        auth_hook_params.body = nil
+    end
+    return auth_hook_params, url
+end
+
+-- timeout in ms
+local function http_req(url, auth_hook_params)
+
+    local httpc = http.new()
+    httpc:set_timeout(1000 * 10)
+    local res, err = httpc:request_uri(url, auth_hook_params)
+    if err then
+        core.log.error("FAIL REQUEST [ ", core.json.encode(auth_hook_params), " ] failed! res is nil, err:", err)
+        return nil, err
+    end
+
+    return res
+end
+
+local function get_auth_info(config, ctx, action, path, client_ip, auth_token)
+    local errmsg
+    local myheaders = request_headers(config, ctx)
+    myheaders["X-Client-Ip"] = client_ip
+    myheaders["Authorization"] = auth_token
+    myheaders["Auth-Hook-Id"] = config.auth_hook_id
+    local args = get_hook_args(config.hook_args)
+    args['hook_path'] = path
+    args['hook_action'] = action
+    args['hook_client_ip'] = client_ip
+    core.response.set_header("hook-cache", 'no-cache')
+    local auth_hook_params, url = hook_configure_params(args, config, myheaders)
+    local res, err = http_req(url, auth_hook_params)
+    if err then
+        core.log.error("fail request: ", url, ", err:", err)
+        return {
+            status = 500,
+            err = "request to hook-server failed, err:" .. err
+        }
+    end
+
+    if res.status ~= 200 and res.status ~= 401 then
+        return {
+            status = 500,
+            err = 'request to hook-server failed, status:' .. res.status
+        }
+    end
+
+    local res_body, res_err = json.decode(res.body)
+    if res_err then
+        errmsg = 'check permission failed! parse response json failed!'
+        core.log.error("json.decode(", res.body, ") failed! err:", res_err)
+        return { status = res.status, err = errmsg }
+    else
+        errmsg = res_body.message
+        return { status = res.status, err = errmsg, body = res_body }
+    end
+end
+
+function _M.rewrite(conf, ctx)
+
+    local url = ctx.var.uri
+    local action = ctx.var.request_method
+    local client_ip = ctx.var.http_x_real_ip or core.request.get_ip(ctx)
+    local config = conf
+    local auth_token, err = get_auth_token(ctx)
+    res_init_headers(config)
+    if auth_token then
+        local res
+        if config.hook_cache then
```

Comment:
I want to know, if we need to cache many requests, for example, in order to reduce frequent requests to the auth-server for the same token, can I use this core.lrucache cache?

Reply:
I am not sure if it is a good idea, since the token may contain client info, which is various.
But since this option is off by default, I can accept it.